pkg/logs/read.go

```diff
@@ -0,0 +1,103 @@
+package logs
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/observatorium/up/pkg/auth"
+	"github.com/observatorium/up/pkg/options"
+	"github.com/observatorium/up/pkg/transport"
+	"github.com/observatorium/up/pkg/util"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+func Read(
+	ctx context.Context,
+	endpoint *url.URL,
+	tp auth.TokenProvider,
+	labels []prompb.Label, // change to Loki ProtoBufs
+	ago, latency time.Duration,
+	m util.Metrics,
+	l log.Logger,
+	tls options.TLS,
+) error {
+	var (
+		rt  http.RoundTripper
+		err error
+	)
+
+	if endpoint.Scheme == transport.HTTPS {
+		rt, err = transport.NewTLSTransport(l, tls)
+		if err != nil {
+			return errors.Wrap(err, "create round tripper")
+		}
+	} else {
+		rt = http.DefaultTransport
+	}
+
+	client := &http.Client{Transport: rt}
+
+	labelSelectors := make([]string, len(labels))
+	for i, label := range labels {
+		labelSelectors[i] = fmt.Sprintf(`%s="%s"`, label.Name, label.Value)
+	}
+
+	query := fmt.Sprintf("{%s}", strings.Join(labelSelectors, ","))
+
+	params := url.Values{}
+	params.Add("query", query)
+	endpoint.RawQuery = params.Encode()
+
+	req, err := http.NewRequest("Get", endpoint.String(), nil)
+	if err != nil {
+		return errors.Wrap(err, "creating request")
+	}
+
+	token, err := tp.Get()
+	if err != nil {
+		return errors.Wrap(err, "retrieving token")
+	}
+
+	if token != "" {
+		req.Header.Add("Authorization", "Bearer "+token)
```

Comment:
rather than re-implement this everywhere, it would be nice to re-use the `newInstantQueryRoundTripper` from the metrics package. Let's put this in the auth package and make it an `AuthorizationTokenRoundTripper` and use it everywhere. WDYT?